wallet-new/src/Cardano/Wallet/API/V1/Transactions.hs

```diff
@@ -0,0 +1,19 @@
+
+module Cardano.Wallet.API.V1.Transactions where
+
+import           Cardano.Wallet.API.V1.Parameters
+import           Cardano.Wallet.API.V1.Types
+
+import           Servant
+
+type API
+    =    "transactions" :> Summary "Generates a new transaction from the source to one or multiple target addresses."
+                        :> ReqBody '[JSON] Payment
+                        :> Post '[JSON] Transaction
+    :<|> "transactions" :> Summary "Returns the transaction history, i.e the list of all the past transactions."
+                        :> WalletRequestParams
+                        :> Get '[JSON] (OneOf [Transaction] (ExtendedResponse [Transaction]))
+    :<|> "transactions" :> "fees"
+                        :> Summary "Estimate the fees which would originate from the payment."
+                        :> ReqBody '[JSON] Payment
+                        :> Post '[JSON] EstimatedFees
```

Comment:
I remember how we discussed a good idea of having hierarchical URI structure for such api endpoints (i. e. `/api/v1/transactions/<wallet_id>/<account_id>/<address_id>`). I don't remember that we ever dismissed that idea. What is your plan on this? Are you going to implement such design later? I personally don't think that we should implement something that we know will have to be re-implemented later.

Reply:
Let me explain the reasoning behind this.

As written [in this RESTful best-practices guide](http://www.restapitutorial.com/media/RESTful_Best_Practices-v1_1.pdf), API are for consumers and the API should make sense to them before everything else, or we have lost in our attempt to make their lives easier.

In the section labeled "Resource URI Examples", there are a bunch of examples for hierarchical resources, and you see  how deeply hierarchical resources would need to be properly interspersed with piece of URLs. For example:

`GET http://www.example.com/customers/33245/orders/8769/lineitems`

Now, this is a properly designed piece of API. You can immediately establish that we are getting all the `lineitems` for order `8769` of customer `33245`. this is because who wrote this enforced the well-known convention of putting free-variables in the URL which are qualified entirely by the URL path which precedes them.

In other terms, here `33245` is immediately after `customers` therefore it must be a customer id. Same goes for `8769`, which follows `orders`.

Now, look at this:

`/api/v1/transactions/J7rQqaLLHBFPrgJXwpktaMB1B1kQBXAyc2uRSfRPzNVGiv6TdxBzkPNBUWysZZZdhFG9gRy3sQFfX5wfpLbi4XTFGFxTg/10/5oP9ib6ym3XdFXvmLXfUVZWw537vicy11ppa5bo8Ku9weByZAUaXWknjiA4dzK84Si`

First of all, this is really hard for a human to parse and visualise, which is not great for ergonomics. Secondly, this betrays the principle above. What's `J7rQqaLLHBFPrgJXwpktaMB1B1kQBXAyc2uRSfRPzNVGiv6TdxBzkPNBUWysZZZdhFG9gRy3sQFfX5wfpLbi4XTFGFxTg`? Is this a transaction ID, a wallet ID, something different? There is no way to tell and you cannot use conventions because this way we have just violated those.

In short, I think hierarchical resources makes lot of sense, if applicable. Here the resource we are dealing with is a `Transaction`. _Transaction and wallet does not form a hierarchy because transaction does not logically nests wallets_. If we had some hierarchical nesting, then it would have made lot of sense to have something like `/api/v1/transactions/<wallet_id>/accounts/<account_id>/addresses<address_id>`, but I think that for this one we should go with my proposal.

I hope this clarifies. What do you think?

Reply:
>Transaction and wallet does not form a hierarchy because transaction does not logically nests wallets.
I'm not sure if I understood your point here. I personally think that if we want to give a user an opportunity to get transaction history not only for a whole wallet, but also for a _particular account_ or a _particular address_, then hierarchical URI make a lot of sense. And we probably do want to give such opportunity (we actually have it in current API, though its implementation is far from perfect). I agree though that we should include `accounts` and `addresses` parts in URI for clarity.

Even if you think that we don't want to provide such opportunity, we should have wallet ID included in the URI anyway (now it looks like the endpoint returns transaction history for _all_ the wallets mixed together, that is totally not what we want :wink:)

Regarding POST requests - this is a bit less intuitive. My vision is that using hierarchical URIs we can provide an option to more finely restrict allowed money source for a new transaction, i. e. a user can use money from the whole wallet, only from single account or only from single address. Though I admit that such URI structure for _creating_ a transaction looks less intuitive and that we probably don't really need to provide such options for user. But if we eventually go with hierarchical URIs for GET requests (getting transaction history), same URI structure for POST requests will look nice and symmetric.

Reply:
I have reflected on what you said, and I see where you are coming from, trust me, but let me explain why I think might be cleaner to go with `/transactions` only 😉 

What you describe are all _filtering_ problems, for which we have an epic [here](https://iohk.myjetbrains.com/youtrack/issue/CSL-1905) . Given a collection, you want to zoom into a subset of elements satisfying some criteria. 

Obviously here you have two choices; you can either nest everything under `/wallets` (like, for example `/wallets/{wallet_id}/accounts/{account_id}/addresses/{address}/transactions`) or to keep everything separated into a separate resource and filter accordingly (e.g. `/transactions?wallet_id=xxx&account_id=xxx&address=xxx`). The first, by experience, comes with some drawbacks:

- It doesn't fully reflect an hierarchy. Typically RESTful resources, when nested, indicates _"a collection of"_. The very same definition of `hierarchy`, in the dictionary, says: _"an arrangement or classification of things according to relative importance or inclusiveness._". Here you don't have the concept of inclusiveness, because a wallet is conceptually composed of accounts & addresses, but _not_ by transactions. Transactions represents separate resources in the system and as such should be separated at the RESTful level.

- It's not composable. In this particular case does not matter, because you cannot further "slice & dice" the data, but you cannot add more filters without breaking the API, which is something you can do if you filter via query params.

Said that, I think for the trifecta `<wallets, accounts, addresses>` it *does* make sense to have the nested representation, I just think transaction shouldn't belong there. If we would follow an arbitrary nesting strategy, you wouldn't know where to stop with nesting. By the same principle, you might in future want to go with:

```
/wallets/{wallet_id}/accounts/{account_id}/addresses/{address_id}/transactions/{transaction_id}/...
```

...and you wouldn't know where to draw a line. The line I'm proposing here is very simple: if you can think about a URL as representing a _view_ or a _zoom_ on a collection, then you should nest the resources, otherwise represent the resources separately.

Said that, considering what's written [here](http://www.restapitutorial.com/lessons/restquicktips.html):

> Producing a great API is 80% art and 20% science. 

and provided frontend devs would also need to use and be comfortable with this API, I suggest maybe @darko-mijic gives us his unbiased opinion on what makes most sense, by experience 😀

Reply:
I fully agree with @adinapoli-iohk here, for us it's only easy to construct URL paths for nested resources (because the path itself never changes, only the provided IDs -> e.g: `/api/v1/transactions/<wallet_id>/accounts/<account_id>/addresses<address_id>`).

For (optional/mixed) filtering params, it's easier to move them into query params, since we can hand them over as a simple object with any combination of properties (which are then turned into query params by the request).

Also for POST requests i fully agree with @adinapoli-iohk that we absolutely have to send the wallet addreses etc. as body params and not include them in the URL (we even had an URL encoding bug because of this in Daedalus / cardano … which was unnecessary).

Standard REST: "you want to create a transaction? Then it's `POST /api/transactions`

Reply:
I fully agree with @DominikGuzei and @adinapoli-iohk.

Reply:
Thanks for chiming in, guys! Ok, looks like we were able to take an informative decision. I'm very happy therefore to stick with the proposed design and review this later in case we find out it's not suitable for our endeavours, but at least we get the prototype rolling and we don't get stuck on design decisions we can review before freezing V1 in place.